File: content/tracks/dsa-typescript/concepts/14-greedy-algorithms/greedy-algorithms_test.ts
import { canJump, mergeIntervals, taskScheduler } from "./stub";

let passed = 0;
let failed = 0;

function test(name: string, fn: () => void): void {
  try {
    fn();
    console.log(`  PASS: ${name}`);
    passed++;
  } catch (e: unknown) {
    const msg = e instanceof Error ? e.message : String(e);
    console.log(`  FAIL: ${name} — ${msg}`);
    failed++;
  }
}

function assertEqual<T>(actual: T, expected: T): void {
  const a = JSON.stringify(actual);
  const b = JSON.stringify(expected);
  if (a !== b) throw new Error(`expected ${b}, got ${a}`);
}

// --- canJump ---

test("canJump: can reach", () => {
  assertEqual(canJump([2, 3, 1, 1, 4]), true);
});

test("canJump: cannot reach", () => {
  assertEqual(canJump([3, 2, 1, 0, 4]), false);
});

test("canJump: single element", () => {
  assertEqual(canJump([0]), true);
});

test("canJump: all zeros except first — blocked", () => {
  assertEqual(canJump([1, 0, 0]), false);
});

test("canJump: large first jump", () => {
  assertEqual(canJump([5, 0, 0, 0, 0, 0]), true);
});

test("canJump: jump exactly to last", () => {
  assertEqual(canJump([2, 0, 0]), true);
});

test("canJump: zero start with two elements", () => {
  assertEqual(canJump([0, 1]), false);
});

// --- mergeIntervals ---

test("mergeIntervals: standard case", () => {
  assertEqual(
    mergeIntervals([[1, 3], [2, 6], [8, 10], [15, 18]]),
    [[1, 6], [8, 10], [15, 18]]
  );
});

test("mergeIntervals: touching intervals merge", () => {
  assertEqual(mergeIntervals([[1, 4], [4, 5]]), [[1, 5]]);
});

test("mergeIntervals: no overlap", () => {
  assertEqual(mergeIntervals([[1, 2], [3, 4]]), [[1, 2], [3, 4]]);
});

test("mergeIntervals: fully contained", () => {
  assertEqual(mergeIntervals([[1, 10], [2, 5]]), [[1, 10]]);
});

test("mergeIntervals: single interval", () => {
  assertEqual(mergeIntervals([[3, 7]]), [[3, 7]]);
});

test("mergeIntervals: all merge into one", () => {
  assertEqual(mergeIntervals([[1, 4], [2, 5], [3, 6]]), [[1, 6]]);
});

test("mergeIntervals: unsorted input", () => {
  assertEqual(
    mergeIntervals([[8, 10], [1, 3], [2, 6], [15, 18]]),
    [[1, 6], [8, 10], [15, 18]]
  );
});

// --- taskScheduler ---

test("taskScheduler: n=2, symmetric tasks", () => {
  assertEqual(taskScheduler(["A","A","A","B","B","B"], 2), 8);
});

test("taskScheduler: n=0, no cooldown", () => {
  assertEqual(taskScheduler(["A","A","A","B","B","B"], 0), 6);
});

test("taskScheduler: single task type", () => {
  // A,idle,idle,A,idle,idle,A
  assertEqual(taskScheduler(["A","A","A"], 2), 7);
});

test("taskScheduler: many task types fill gaps", () => {
  // Enough variety that no idling needed
  assertEqual(taskScheduler(["A","A","A","B","B","C","C","D","E"], 2), 9);
});

test("taskScheduler: n=1", () => {
  assertEqual(taskScheduler(["A","A","A","B","B","B"], 1), 6);
});

test("taskScheduler: single task", () => {
  assertEqual(taskScheduler(["A"], 5), 1);
});

console.log(`\n${passed} passed, ${failed} failed`);
if (failed > 0) process.exit(1);

File: content/tracks/dsa-python/concepts/14-greedy-algorithms/greedy-algorithms_test.py
import sys
from stub import can_jump, merge_intervals, task_scheduler

passed = 0
failed = 0


def test(name: str, condition: bool, detail: str = "") -> None:
    global passed, failed
    if condition:
        print(f"  PASS: {name}")
        passed += 1
    else:
        msg = f" — {detail}" if detail else ""
        print(f"  FAIL: {name}{msg}")
        failed += 1


# can_jump
test("jump: reachable [2,3,1,1,4]", can_jump([2, 3, 1, 1, 4]) is True)
test("jump: stuck at 0 [3,2,1,0,4]", can_jump([3, 2, 1, 0, 4]) is False)
test("jump: single element [0]", can_jump([0]) is True)
test("jump: two elements reachable [1,0]", can_jump([1, 0]) is True)
test("jump: stuck at first [0,1]", can_jump([0, 1]) is False)
test("jump: jump over zeros [2,0,0]", can_jump([2, 0, 0]) is True)
test("jump: all ones", can_jump([1, 1, 1, 1, 1]) is True)

# merge_intervals
r = merge_intervals([[1, 3], [2, 6], [8, 10], [15, 18]])
test("merge: standard", r == [[1, 6], [8, 10], [15, 18]], f"got {r}")
r = merge_intervals([[1, 4], [4, 5]])
test("merge: touching intervals", r == [[1, 5]], f"got {r}")
r = merge_intervals([[1, 4], [2, 3]])
test("merge: contained interval", r == [[1, 4]], f"got {r}")
r = merge_intervals([])
test("merge: empty", r == [], f"got {r}")
r = merge_intervals([[1, 4]])
test("merge: single interval", r == [[1, 4]], f"got {r}")
r = merge_intervals([[1, 2], [3, 4], [5, 6]])
test("merge: no overlaps", r == [[1, 2], [3, 4], [5, 6]], f"got {r}")
r = merge_intervals([[2, 3], [1, 4]])
test("merge: unsorted input", r == [[1, 4]], f"got {r}")
r = merge_intervals([[1, 10], [2, 3], [4, 5]])
test("merge: one contains all", r == [[1, 10]], f"got {r}")

# task_scheduler
r = task_scheduler(["A","A","A","B","B","B"], 2)
test("scheduler: n=2 (expect 8)", r == 8, f"got {r}")
r = task_scheduler(["A","A","A","B","B","B"], 0)
test("scheduler: n=0 (expect 6)", r == 6, f"got {r}")
r = task_scheduler(["A","A","A","A","A","A","B","C","D","E","F","G"], 2)
test("scheduler: many tasks n=2 (expect 16)", r == 16, f"got {r}")
r = task_scheduler(["A"], 1)
test("scheduler: single task n=1 (expect 1)", r == 1, f"got {r}")
r = task_scheduler(["A","A","A","B","B","B","C","C","C"], 2)
test("scheduler: three equal freq n=2 (expect 9)", r == 9, f"got {r}")

print(f"\n{passed} passed, {failed} failed")
if failed > 0:
    sys.exit(1)
